webrender/src/device.rs

```diff
@@ -770,23 +748,13 @@ impl Device {
 
     pub fn compile_shader(gl: &gl::Gl,
                           name: &str,
-                          source_str: &str,
                           shader_type: gl::GLenum,
-                          shader_preamble: &[String])
+                          sources: &[String])
                           -> Result<gl::GLuint, ShaderError> {
         debug!("compile {:?}", name);
-
-        let mut s = String::new();
-        s.push_str(get_shader_version(gl));
-        for prefix in shader_preamble {
-            s.push_str(prefix);
-        }
-        s.push_str(source_str);
-
         let id = gl.create_shader(shader_type);
-        let mut source = Vec::new();
-        source.extend_from_slice(s.as_bytes());
-        gl.shader_source(id, &[&source[..]]);
+        let sources: Vec<&[u8]> = sources.iter().map(|s| s.as_bytes()).collect();
```

Comment:
I'd prefer us building a single source chunk per shader (as opposed to having an array of chunks), because OpenGL is the only API supporting multiple sources.